server/http/helloapp/locations.go

```diff
@@ -0,0 +1,112 @@
+package helloapp
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/mattermost/mattermost-plugin-apps/server/apps"
+	"github.com/mattermost/mattermost-plugin-apps/server/utils/httputils"
+	"github.com/pkg/errors"
+)
+
+func CheckAuthentication(f func(http.ResponseWriter, *http.Request)) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, req *http.Request) {
+		authValue := req.Header.Get(apps.OutgoingAuthHeader)
+		if !strings.HasPrefix(authValue, "Bearer ") {
+			httputils.WriteBadRequestError(w, errors.Errorf("missing %s: Bearer header", apps.OutgoingAuthHeader))
+			return
+		}
+
+		jwtoken := strings.TrimPrefix(authValue, "Bearer ")
+		claims := apps.JWTClaims{}
+		_, err := jwt.ParseWithClaims(jwtoken, &claims, func(token *jwt.Token) (interface{}, error) {
+			if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+				return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
+			}
+			return []byte(AppSecret), nil
+		})
+		if err != nil {
+			httputils.WriteBadRequestError(w, err)
+			return
+		}
+
+		f(w, req)
+	}
+}
+
+func ExtractUserAndChannelID(f func(http.ResponseWriter, *http.Request, string, string)) func(http.ResponseWriter, *http.Request) {
```

Comment:
Can we move this function and `CheckAuthentication` into a different file(s)?

Could we actually place this code in `mattermost-plugin-apps/server/apps`, to be shareable to other Go programs? And to have the parsing logic right by the marshaling logic.

Reply:
Since these are App related, I am not sure whether we should move them back to apps. Any app can be written in any language, not just Go, and they may or may not use the same pattern as we do. Therefore might not be useful.

Also, I am 0/5 on whether a Go app should import this repo to access those utils. Maybe a different utils repo? Something like mattermost-plugins-apps-utils? Nevertheless, I know there is some thoughts against utils repos :sweat_smile:

Reply:
It seems unorthodox to do what I suggested, but I still think it's not a bad idea. Of course apps can be written in other languages, but I think it's fine for a plugin or app written in go to import the `mattermost-plugin-apps` repo, just as we import `mattermost-server` in the plugins. Different scenario of course, but I still like how both sides of the JWT conversation would be right next to each other. We are already importing things from the `apps` package in the `helloapp` package, so I think it's not a big deal here.